packages/flutter/test/widgets/navigator_replacement_test.dart

```diff
@@ -39,42 +42,202 @@ void main() {
     expect(find.text('home'), findsOneWidget);
   });
 
-  testWidgets('pushAndRemoveUntil', (WidgetTester tester) async {
-    await tester.pumpWidget(MaterialApp(
+  group('pushAndRemoveUntil', () {
+
+    final TestObserver observer = TestObserver();
+    final Widget myApp = MaterialApp(
       home: const Material(child: Text('home')),
       routes: <String, WidgetBuilder>{
         '/a': (BuildContext context) => const Material(child: Text('a')),
         '/b': (BuildContext context) => const Material(child: Text('b')),
+        '/c': (BuildContext context) => const Material(child: Text('c')),
       },
-    ));
-
-    final NavigatorState navigator = tester.state(find.byType(Navigator));
-    navigator.pushNamed('/a');
-    await tester.pumpAndSettle();
-
-    expect(find.text('home', skipOffstage: false), findsOneWidget);
-    expect(find.text('a', skipOffstage: false), findsOneWidget);
-    expect(find.text('b', skipOffstage: false), findsNothing);
-
-    navigator.pushNamedAndRemoveUntil('/b', (Route<dynamic> route) => false);
-    await tester.pumpAndSettle();
-
-    expect(find.text('home', skipOffstage: false), findsNothing);
-    expect(find.text('a', skipOffstage: false), findsNothing);
-    expect(find.text('b', skipOffstage: false), findsOneWidget);
-
-    navigator.pushNamed('/');
-    await tester.pumpAndSettle();
-
-    expect(find.text('home', skipOffstage: false), findsOneWidget);
-    expect(find.text('a', skipOffstage: false), findsNothing);
-    expect(find.text('b', skipOffstage: false), findsOneWidget);
-
-    navigator.pushNamedAndRemoveUntil('/a', ModalRoute.withName('/b'));
-    await tester.pumpAndSettle();
-
-    expect(find.text('home', skipOffstage: false), findsNothing);
-    expect(find.text('a', skipOffstage: false), findsOneWidget);
-    expect(find.text('b', skipOffstage: false), findsOneWidget);
+      navigatorObservers: <NavigatorObserver>[observer],
+    );
+
+    testWidgets('notifies appropriately', (WidgetTester tester) async {
+
+      await tester.pumpWidget(myApp);
+
+      final NavigatorState navigator = tester.state(find.byType(Navigator));
+      navigator.pushNamed('/a');
+      await tester.pumpAndSettle();
+
+      expect(find.text('home', skipOffstage: false), findsOneWidget);
+      expect(find.text('a', skipOffstage: false), findsOneWidget);
+      expect(find.text('b', skipOffstage: false), findsNothing);
+
+      bool isPushed = false;
```

Comment:
This test may be a little easier to understand by just having a `List<String> log = <String>[]` and on Pushed, onRemoved just add strings like `/a pushed, previousRoute: /` or `/ removed, previousRoute: /a` to that log. You can then assert that the log contains the expected messages.